fem/tmop.cpp

```diff
@@ -1363,6 +1554,207 @@ void TMOP_Integrator::AssembleElementGrad(const FiniteElement &el,
+                               const MPI_Comm &comm)
+{
+   if (fdflag==0) {return 0.;}
+   double min_jac = ComputeMinJac(x,fes);
+   double min_jac_all;
+   MPI_Allreduce(&min_jac, &min_jac_all, 1, MPI_DOUBLE, MPI_MIN, comm);
+   fdeps = pow(10,-2.)*min_jac_all;
+   return fdeps;
+}
+#endif
+
+double TMOP_Integrator::GetFDDerivative(const FiniteElement &el,
+                                        ElementTransformation &T,
+                                        Vector &elfun,
+                                        const int nodenum,const int idir)
+{
+   int dof = el.GetDof();
+   int idx = idir*dof+nodenum;
+   elfun[idx]    += fdeps;
+   double energy1 = GetElementEnergy(el, T, elfun);
+   elfun[idx]    -= fdeps;
+   double energy2 = elemenergy;
+   double der     = (energy1-energy2)/fdeps;
+
+   (*(ElemPertEnergy[T.ElementNo]))(idx) = energy1;
+   (*(ElemDer[T.ElementNo]))(idx) = der;
+
+   return der;
+}
+
+void TMOP_Integrator::AssembleElementVectorFD(const FiniteElement &el,
+                                              ElementTransformation &T,
+                                              const Vector &elfun, Vector &elvect)
+{
+   const int dof = el.GetDof(), dim = el.GetDim();
+   int elnum = T.ElementNo;
+   if (elnum>=ElemDer.Size())
+   {
+      ElemDer.Append(new Vector);
+      ElemPertEnergy.Append(new Vector);
+   }
+   if (ElemDer[elnum]->Size() != dof*dim)
+   {
+      ElemDer[elnum]->SetSize(dof*dim);
+      ElemPertEnergy[elnum]->SetSize(dof*dim);
+   }
+
+   DSh.SetSize(dof, dim);
+   DS.SetSize(dof, dim);
+   Jrt.SetSize(dim);
+   Jpt.SetSize(dim);
+   P.SetSize(dim);
+   PMatI.UseExternalData(elfun.GetData(), dof, dim);
+   elvect.SetSize(dof*dim);
+   Vector elfunmod;
+   elfunmod.SetSize(dof*dim);
+   elfunmod.SetData(elfun.GetData());
+
+   // Energy for unperturbed configuration
+   elemenergy = GetElementEnergy(el, T, elfun);
+
+   for (int j=0; j<dim; j++)
+   {
+      for (int i=0; i<dof; i++)
+      {
+         if (discr_tc)
+         {
+            discr_tc->UpdateTargetSpecificationAtNode(
+               el,T,i,j,discr_tc->tspec_perth);
+         }
+         elvect(j*dof+i) = GetFDDerivative(el,T,elfunmod,i,j);
+         if (discr_tc) {discr_tc->RestoreTargetSpecificationAtNode(T,i);}
+      }
+   }
+}
+
+void TMOP_Integrator::AssembleElementGradFD(const FiniteElement &el,
+                                            ElementTransformation &T,
+                                            const Vector &elfun,
+                                            DenseMatrix &elmat)
+{
+   const int dof = el.GetDof(), dim = el.GetDim();
+   DSh.SetSize(dof, dim);
+   DS.SetSize(dof, dim);
+   Jrt.SetSize(dim);
+   Jpt.SetSize(dim);
+   PMatI.UseExternalData(elfun.GetData(), dof, dim);
+   elmat.SetSize(dof*dim);
+
+   Vector elfunmod;
+   elfunmod.SetSize(dof*dim);
+   elfunmod.SetData(elfun.GetData());
+
+   Vector ElemDerLoc = *(ElemDer[T.ElementNo]);
```

Comment:
I guess this assumes that `AssembleElementVectorFD` has been called previously on the same mesh (maybe add a comment?) You should also do references instead of copies.